doc/developer/design/20240205_cluster_specific_optimization.md

```diff
@@ -0,0 +1,283 @@
+2.  For each such cluster, run the following experiment:
+
+The suggested workflow for running an experiment on a specific customer
+environment will be:
+
+1.  Obtain `mz_system` privileges through Teleport.
+2.  Create a dedicated cluster for the experiment. Use the feature flag to
+    enable the tested feature in the `CREATE CLUSTER` definition.
+3.  Create an `UNBILLED` replica for that cluster.
+4.  Ask the customer to replicate (a subset of) dataflow-backed catalog items
+    defined on the the original cluster to the experiment cluster.
+5.  Monitor and record observed differences between the dataflows running in the
+    original cluster and the (modified) dataflows running in the experiment
+    cluster.
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+In order to facilitate this workflow, we propose the the following changes
+(discussed in detail below):
+
+-   Extensions to the `CREATE CLUSTER` syntax.
+-   Extensions to the optimizer API.
+
+### Extensions to the `CREATE CLUSTER` syntax
+
+Extend the `CREATE CLUSTER SYNTAX` for managed cluster plans as follows:
+
+```sql
+CREATE CLUSTER cluster
+  REPLICAS (
+    ...
+  )
+  WITH FEATURES (
+    enable_eager_delta_joins = true
+  )
+```
+
+The newly added `WITH FEATURES` clause will be only visible by `mz_system`. We
+will extend `CreateClusterManagedPlan` and `ClusterVariantManaged` with a new
+`ClusterFeatures` struct that models the feature flag configuration that can be
+attached on a per cluster level.
+
+### Extensions to the `optimize` API
+
+Here we can benefit from the unified optimizer interface introduced with
+MaterializeInc/materialize#20569. As part of the associated changes we
+introduced an `mz_adapter::optimize::OptimizerConfig` struct that currently can
+can already be configured in a layered way:
+
+1. A top layer of settings bound from `SystemVars`.
+2. A layer of `EXPLAIN`-specific overrides bound from the `ExplainContext`,
+   which is available when the `Coordinator` methods that drive the optimization
+   process are initiated from `sequence_explain_plan`.
+
+Since all `Optimizer::new(...)` constructor calls in the `Coordinator` happen at
+a time where the target cluster for the optimized statement is already resolved,
+we can just add a new layer for cluster-specific overrides between (1) and (2).
```

Comment:
Will we still want `OptimizerConfig` to have a bunch of individual flags, or do we want to store the `WITH FEATURES` overrides some other way?

Reply:
Here is the current thinking:

`OptimizerConfig` will declare optimizer flags as fields.

Per default, the values of these flags will be bound from the system-wide `SystemVars` (which is synced from LaunchDarky).

Currently, there is an override supported for `ExplainConfig` where certain flags are defined as `flag_name: Option<flag_type>`. Whenever you type `EXPLAIN WITH(flag_name)` the value of this `ExplainConfig` field is populated, and in the `override_from<ExplainConfig>` implementation a present value is used to override the current value.

https://github.com/MaterializeInc/materialize/blob/c1a15827223d5c11e2ec447672bc81c684f594bd/src/adapter/src/optimize/mod.rs#L214-L232

The `Coordinator` constructs the config for all optimizer instances like that:

https://github.com/MaterializeInc/materialize/blob/de7b7c87e6bf482393c0542abc31be6c32f456ef/src/adapter/src/coord/sequencer/inner/peek.rs#L288-L289

The plan is to extend the `ClusterVariantManaged` with a dedicated struct (`ClusterFeatures` or something similar) that also has `flag_name: Option<flag_type>` fields and change the `optimizer_config` init code to

```rust
let optimizer_config = optimize::OptimizerConfig::from(self.catalog().system_config())
    .override_from(&cluster_features)
    .override_from(&explain_ctx);
```

Then we should ensure that parts of the optimizer pipeline that can be parameterized are called with parameters bound from the associated `OptimizerConfig` instance. I have a convention in mind for how to minimize amount of code changes per feature flag needed to do facilitate that, but need to sketch this before I can say more.

Reply:
It seems like we'll also want to adjust the way we define feature flags to indicate which feature flags are usable in a `CREATE CLUSTER ... FEATURES` clause. Do you have a sense of how that will work?

Reply:
This is now visible in #25373. Sadly it wasn't so easy to integrate the definition ones. For the MVP I was able to reduce the amount of changes to the following five steps (which will be added to the `mz_adapter::optimize` rustdoc):

1. Add the feature flag as a field in the `OptimizerConfig` definition in `mz_adapter::optimize`. This makes the flag available to stages in our optimizer pipeline.
2. Define the feature flag in the `feature_flags!(...)` call in `mz_sql::session::vars` and extend the `impl From<&SystemVars> for OptimizerConfig` definition in `mz_adapter::optimize`. This will allow uses to set a system-wide override for this feature flag.
3. For the `EXPLAIN ... WITH(...)` statement:
    1. Add the feature flag to the `impl_simple_options!(ExplainPlanOption { ... })` call.
    2. Extend the `impl TryFrom<ExplainPlanOptionExtracted> for ExplainConfig` definition in `mz_sql::plan::statement::dml`.
    3. Extend the `OverrideFrom<ExplainContext>` implementation in `mz_adapter::optimize`.
4. For the `CLUSTER ... FEATURES(...)` statement:
    1. Add the feature flag to the `impl_simple_options!(ClusterFeature { ... })` call.
    2. Extend the `let optimizer_feature_overrides = OptimizerFeatureOverrides { ... }` call in `mz_sql::plan::statement::ddl::plan_create_cluster`.
    3. Define the feature flag in the `optimizer_feature_flags!` macro in `mz_repr::optimize`.
    4. Extend the `OverrideFrom<OptimizerFeatureOverrides>` implementation in `mz_adapter::optimize`.

As you can see the number of steps is non-trivial. However, with the addition of the `feature_flags!` and `impl_simple_options!` macros in #25373 at least the boilerplate code that needs to be maintained is reduced—each step requires changing 1-3 LOC per feature flag. I think also I can get rid of the steps 4.iv and 3.iii.